### PR TITLE
[Rust] ObjectStore::read_dir method

### PR DESCRIPTION
**Pull request by eddyxu**
To support scan directory on local and cloud storage.

### CODE REVIEW

**Comment by eddyxu** [comment]
Will add cloud storage tests into integration tests.

File: rust/src/io/object_store.rs [hunk 1]
```diff
@@ -176,6 +176,20 @@ impl ObjectStore {
         ObjectWriter::new(self, path).await
     }
 
+    /// Read a directory (start from base directory) and returns all sub-paths in the directory.
+    pub async fn read_dir(&self, dir_path: impl Into<Path>) -> Result<Vec<String>> {
```
Comment (changhiskhan) [inline rust/src/io/object_store.rs:180]: Should this return `Result<Vec<Path>>` instead for downstream processing?

File: rust/src/io/object_store.rs [hunk 2]
```diff
@@ -176,6 +176,20 @@ impl ObjectStore {
...
+        let path = dir_path.into();
+        let path = Path::parse(path.to_string())?;
+        let output = self.inner.list_with_delimiter(Some(&path)).await?;
+        output
+            .common_prefixes
+            .iter()
+            .map(|cp| cp.filename().map(|s| s.to_string()).unwrap_or_default())
+            .chain(output.objects.iter().map(|o| o.location.to_string()))
+            .map(|s| Ok(Path::parse(s)?.filename().unwrap().to_string()))
```
Comment (changhiskhan) [inline rust/src/io/object_store.rs:189]: is this returning just the shortname or the full path? At the object store level, maybe it makes more sense to get the full paths? and then the lancedb layer can take just the shortname before the .lance as the table name?

**Review by changhiskhan — COMMENTED** [review]
Should this returna

**Comment by changhiskhan** [comment]
also windows failure

File: rust/src/io/object_store.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/io/object_store.rs:189]: This returns the filename, which is the same behavior as the OS `readdir()` call. LanceDB uses the `readdir()` semantic, right? For example, to list what tables are available.

File: rust/src/io/object_store.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/io/object_store.rs:180]: Ok, will fix it.

**Review by changhiskhan — APPROVED** [review]
if we're returning just the filename, does it make sense anymore to return Path?

**Comment by eddyxu** [comment]
> if we're returning just the filename, does it make sense anymore to return Path?

Change it back to return `String`

**Pull request merged**